### PR TITLE
Add TopN optimization in physical plan mapping

### LINKED ISSUE #11260: Top-N optimization not applied when using CTEs
### What happens?

DuckDB does not apply the Top-N optimization step for a query if the ORDER BY and LIMIT clauses are part of different common table expressions. For example, it is expected that these queries would be evaluated identically:

```
SELECT * FROM tbl ORDER BY col LIMIT N;

WITH cte AS (SELECT * FROM tbl ORDER BY col) SELECT * FROM cte LIMIT N;
```

However, according to `EXPLAIN ANALYZE`, only the first query uses a `TOP_N` step. The second query uses `ORDER_BY + LIMIT` which requires sorting the entire result.

### To Reproduce

The first query takes 0.3s to execute while the second takes almost 12s:

```python
import duckdb
duckdb.execute('EXPLAIN ANALYZE SELECT * FROM range(100000000) ORDER BY range LIMIT 100')
print(duckdb.fetchone()[1])
```
```
┌─────────────────────────────────────┐
│┌───────────────────────────────────┐│
││    Query Profiling Information    ││
│└───────────────────────────────────┘│
└─────────────────────────────────────┘
EXPLAIN ANALYZE SELECT * FROM range(100000000) ORDER BY range LIMIT 100
┌─────────────────────────────────────┐
│┌───────────────────────────────────┐│
││         Total Time: 0.349s        ││
│└───────────────────────────────────┘│
└─────────────────────────────────────┘
┌───────────────────────────┐
│      RESULT_COLLECTOR     │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             0             │
│          (0.00s)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│      EXPLAIN_ANALYZE      │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             0             │
│          (0.00s)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│           TOP_N           │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│          Top 100          │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│    "range"."range" ASC    │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│            100            │
│          (0.30s)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│           RANGE           │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│       EC: 100000000       │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│         100000000         │
│          (0.04s)          │
└───────────────────────────┘
```
```python
duckdb.execute('EXPLAIN ANALYZE WITH cte AS (SELECT * FROM range(100000000) ORDER BY range) SELECT * FROM cte LIMIT 100')
print(duckdb.fetchone()[1])
```
```
┌─────────────────────────────────────┐
│┌───────────────────────────────────┐│
││    Query Profiling Information    ││
│└───────────────────────────────────┘│
└─────────────────────────────────────┘
EXPLAIN ANALYZE WITH cte AS (SELECT * FROM range(100000000) ORDER BY range) SELECT * FROM cte LIMIT 100
┌─────────────────────────────────────┐
│┌───────────────────────────────────┐│
││         Total Time: 11.90s        ││
│└───────────────────────────────────┘│
└─────────────────────────────────────┘
┌───────────────────────────┐
│      RESULT_COLLECTOR     │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             0             │
│          (0.00s)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│      EXPLAIN_ANALYZE      │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             0             │
│          (0.00s)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│           LIMIT           │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│            100            │
│          (0.00s)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│          ORDER_BY         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│          ORDERS:          │
│    "range"."range" ASC    │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│            4096           │
│          (7.76s)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│           RANGE           │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│       EC: 100000000       │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│         100000000         │
│          (0.03s)          │
└───────────────────────────┘

```

### OS:

Linux x86_64

### DuckDB Version:

0.10.1

### DuckDB Client:

Python

### Full Name:

Simin Wang

### Affiliation:

Clarity

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a nightly build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by kryonix**
This PR fixes #11260 by introducing the TopN optimization to the physical plan mapping of `LogicalLimit` operators.

### CODE REVIEW

**Comment by kryonix** [comment]
This should now work as intended, _i.e._, apply the TopN optimization when possible during physical plan mapping.
The principle idea of this PR is that unnecessary projections are eliminated when the physical plan is created. This then allows us to use the TopN operator instead of LIMIT and ORDER BY, when certain conditions are met.

**Review by Tmonster — CHANGES_REQUESTED** [review]
Thanks for the PR!

This will definitely work, however, we like to avoid having these kind of optimizations during the physical planning phase whenever possible. If I'm not mistaken we can move this logic to the the topn optimizer itself. It would require just a bit more work on the PR. If the operator is a limit, you can loop through any child projections until you get to an order by. If you encounter a different operator before then, you don't apply the optimization.

If you could add a test and a micro benchmark that would be great!
See `test/optimizer/topn_optimizer.test` for test examples, 
`benchmark/micro/limit/parallel_limit.benchmark` is a minor benchmark example you can copy and modify. Maybe a different directory though, `benchmark/micro/optimizer/` would be good.

**Comment by kryonix** [comment]
Thanks for the feedback @Tmonster, I will take a look at your suggestions 👍🏻

**Review by Tmonster — CHANGES_REQUESTED** [review]
Thanks for the fixes!

Looks much better now. Still some improvements to be made. Let me know if you need any help

**Comment by Tmonster** [comment]
Also can you rebase with feature and change the base of this PR to merge to feature? 
We are working on release v1.0.0 so only bug fixes are getting merged now. Optimizers are a gray area, but this is probably a feature instead of a bug

File: src/optimizer/topn_optimizer.cpp [hunk 1]
```diff
@@ -38,6 +38,66 @@ unique_ptr<LogicalOperator> TopN::Optimize(unique_ptr<LogicalOperator> op) {
...
+		if (op->type == LogicalOperatorType::LOGICAL_LIMIT) {
+			auto &limit = op->Cast<LogicalLimit>();
+			if (limit.limit_val.Type() == LimitNodeType::CONSTANT_VALUE &&
+			    limit.offset_val.Type() != LimitNodeType::EXPRESSION_VALUE) {
+				auto child_op = op->children[0].get();
+				bool can_optimize = false;
+
+				// Check if there are only projection operators between the limit operator
+				// and an order by operator.
+				while (child_op && !can_optimize) {
```
[the remark this answers is not in the document]
Comment (kryonix) [reply src/optimizer/topn_optimizer.cpp]: Makes sense! I fixed that.

File: test/optimizer/topn_optimizer.test [hunk 2]
```diff
@@ -37,3 +37,9 @@ query II
 EXPLAIN SELECT i FROM integers ORDER BY i OFFSET 3
 ----
 logical_opt	<!REGEX>:.*TOP_N.*
+
+# only limit, but in subquery with projection pull up optimization
+query II
+EXPLAIN SELECT * FROM (SELECT * FROM range(100000000) AS _(x) ORDER BY x) AS cte LIMIT 10
```
[the remark this answers is not in the document]
Comment (kryonix) [reply test/optimizer/topn_optimizer.test:43]: Good catch. I've added the test.

File: src/optimizer/topn_optimizer.cpp [hunk 3]
```diff
@@ -38,6 +38,66 @@ unique_ptr<LogicalOperator> TopN::Optimize(unique_ptr<LogicalOperator> op) {
...
+						child_op = nullptr;
+						can_optimize = false;
+						break;
+					}
+				}
+
+				if (can_optimize) {
+					// traverse operator tree and collect all projection nodes until we reach
+					// the order by operator
+					vector<unique_ptr<LogicalOperator>> projections;
```
[the remark this answers is not in the document]
Comment (kryonix) [reply src/optimizer/topn_optimizer.cpp]: This code no longer exists separately. I've merged it with the existing case.

**Comment by kryonix** [comment]
Thank you for the review @Tmonster, I've added all the changes. I did rebase to feature, too.

File: src/optimizer/topn_optimizer.cpp [hunk 4]
```diff
@@ -20,13 +19,41 @@ bool TopN::CanOptimize(LogicalOperator &op) {
...
 	}
 	return false;
 }
 
 unique_ptr<LogicalOperator> TopN::Optimize(unique_ptr<LogicalOperator> op) {
 	if (CanOptimize(*op)) {
+
+		vector<unique_ptr<LogicalOperator>> projections;
+
+		if (op->children[0]->type != LogicalOperatorType::LOGICAL_ORDER_BY) {
```
Comment (Tmonster) [inline src/optimizer/topn_optimizer.cpp:40]: something like might look nicer. If there is no projection sitting in the way, the while loop won't run. You already know you can optimize, so the cast will be valid (and the `D_ASSERT` will check it in debug mode.
```
auto child = std::move(op->children[0]);
// collect all projections until we get to the order by
while (child->type == LogicalOperatorType::LOGICAL_PROJECTION) {
	D_ASSERT(!child->children.empty());
	auto tmp = std::move(child->children[0]);
	projections.push_back(std::move(child));
	child = std::move(tmp);
}
D_ASSERT(child->type == LogicalOperatorType::LOGICAL_ORDER_BY);
auto &order_by = child->Cast<LogicalOrder>();
```
might look nicer.

**Review by Tmonster — CHANGES_REQUESTED** [review]
Nice! Looks very good now, just 2 more small comments then it's good to go 👍

File: src/optimizer/topn_optimizer.cpp [hunk 5]
```diff
@@ -37,6 +64,14 @@ unique_ptr<LogicalOperator> TopN::Optimize(unique_ptr<LogicalOperator> op) {
...
+
+		// reconstruct all projection nodes above limit operator
+		while (!projections.empty()) {
+			auto node = std::move(projections.back());
+			node->children[0] = std::move(op);
+			op = std::move(node);
+			projections.pop_back();
+		}
 	} else {
 		for (auto &child : op->children) {
```
[the remark this answers is not in the document]
Comment (kryonix) [reply src/optimizer/topn_optimizer.cpp]: Good catch. I did remove that.

File: src/optimizer/topn_optimizer.cpp (hunk 4 above)
Comment (kryonix) [reply src/optimizer/topn_optimizer.cpp]: Awesome, that looks good to me.

**Review by Tmonster — APPROVED** [review]
Looks good to me! Thanks 👍 
Our CI to automatically run tests is a bit broken right now. So to run CI, you need to manually convert it to a draft, then convert it back to say ready to review. You can find the button to it it under the "reviewers"

**Comment by Tmonster** [comment]
Hi @kryonix ,

I think you need to run a `make format-fix` and the Code quality CI will turn green. I don't know why the Linux extensions CI is failing. I think that is just noise.

**Comment by kryonix** [comment]
Hi @Tmonster, I did just that before my last commit. `make format-fix` does not make a difference. I merged with feature just now because there were some file conflicts. Maybe that will fix it.

**Review by Tmonster — APPROVED** [review]
LGTM!

**Pull request merged**